operators/pkg/controller/elasticsearch/mutation/comparison/pod.go

```diff
@@ -59,12 +54,33 @@ func PodMatchesSpec(
 	return true, nil, nil
 }
 
-// getEsContainer returns the elasticsearch container in the given pod
-func getEsContainer(containers []corev1.Container) (corev1.Container, error) {
-	for _, c := range containers {
-		if c.Name == v1alpha1.ElasticsearchContainerName {
-			return c, nil
+// ComparePodSpec returns a ComparisonMatch if the given spec matches the spec of the given pod.
+// Comparison is based on the hash of the pod spec (before resource creation), stored in a label in the pod.
+func ComparePodSpec(spec corev1.PodSpec, existingPod corev1.Pod) Comparison {
+	existingPodHash := hash.GetSpecHashLabel(existingPod.Labels)
+	if existingPodHash == "" {
+		return ComparisonMismatch(fmt.Sprintf("No %s label set on the existing pod", hash.SpecHashLabelName))
+	}
+	if hash.HashObject(spec) != existingPodHash {
+		return ComparisonMismatch("Spec hash and running pod spec hash are not equal")
+	}
+	return ComparisonMatch
+}
+
+// MapSubsetComparison returns ComparisonMatch if the expected labels (keys and values) are present in the
+// actual map. The actual map may contain more entries: this will not cause a mismatch.
+// This allows user to add additional labels or annotations to pods, while not causing the pod to be replaced.
+func MapSubsetComparison(expected map[string]string, actual map[string]string, mismatchMsg string) Comparison {
```

Comment:
I have a concern here because usually with `deployments` or `statefulset` user is allowed to change the value of an annotation _(or remove the annotation)_ without triggering a new deployment.

It may sound weird but it might be used by a mutation webhook or an external system to store some informations on the pods that are running in a K8S cluster. (tbh this is what happen for me, I'm playing with `MutatingWebhook` which updates a annotation)

I don't know if we want to be compatible with that but it is just to point out that we may still have some issues with `MutatingWebhook` or any external system that would add/update/delete some annotations while it is fine with `deployments` or `statefulset`.

Reply:
Good point!
Should we maybe include the initial labels & annotations the pod was created with into the hash computation? A user may update/remove/add any label/annotation on a running pod: since we  would compare with the initially created labels & annotations and not the current ones, we would not trigger a pod replacement, unless we want to change that initial set of labels?